### ISSUE TITLE
Support for private github repos in task update_dependencies

### ISSUE THREAD

**Issue by 1handclapping** [SFDO-Tooling/CumulusCI#793]
When running `cci task run update_dependencies` that include references to private Github repositories, the task fails with error: `zipfile.BadZipfile: File is not a zip file`

I took a look around and it appears that the request made by `download_extract_zip` in `cumulusci.utils` is using only the github auth token in `headers` and this fails when retrieving a private github repo, returning a `Not Found` string. I was able to patch this in my case by adding an auth keyword arg to `download_extract_zip` that is passed onto `requests`. Not sure if that is the best way forward or not or if this is a bug or enhancement.

Thanks for a great tool.

```
 cci task run update_dependencies
Checking the version!
2018-10-01 11:05:12: Getting scratch org info from Salesforce DX
2018-10-01 11:05:17: Beginning task: UpdateDependencies
2018-10-01 11:05:17:        As user: test-xijve254x7dd@example.com
2018-10-01 11:05:17:         In org: 00DN0000000ENrn
2018-10-01 11:05:17:
2018-10-01 11:05:17: Preparing static dependencies map
2018-10-01 11:05:17: Processing dependencies from Github repo https://github.com/1handclapping/fflib-base
2018-10-01 11:05:19: Processing dependencies from Github repo https://github.com/1handclapping/corelogic-sf-api
2018-10-01 11:05:21: Retrieving list of packages from target org
2018-10-01 11:05:21: Pending
2018-10-01 11:05:23: [Done]
2018-10-01 11:05:25: Dependencies:
2018-10-01 11:05:25:     subfolder: fflib-base-master/src
2018-10-01 11:05:25:     unmanaged: True
2018-10-01 11:05:25:     zip_url: https://github.com/1handclapping/fflib-base/archive/master.zip
2018-10-01 11:05:25:     subfolder: corelogic-sf-api-master/src
2018-10-01 11:05:25:     unmanaged: True
2018-10-01 11:05:25:     zip_url: https://github.com/1handclapping/corelogic-sf-api/archive/master.zip
2018-10-01 11:05:25: Deploying unmanaged metadata from /fflib-base-master/src of https://github.com/1handclapping/fflib-base/archive/master.zip
Traceback (most recent call last):
  File "/Users/ant/.pyenv/versions/cci/bin/cci", line 11, in <module>
    load_entry_point('cumulusci', 'console_scripts', 'cci')()
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/lib/python2.7/site-packages/click/decorators.py", line 27, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/cli/cci.py", line 991, in task_run
    handle_exception_debug(config, debug, no_prompt=no_prompt)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/cli/cci.py", line 974, in task_run
    task()
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/core/tasks.py", line 131, in __call__
    self.result = self._run_task()
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/tasks/salesforce/UpdateDependencies.py", line 76, in _run_task
    self._install_dependencies()
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/tasks/salesforce/UpdateDependencies.py", line 174, in _install_dependencies
    self._install_dependency(dependency)
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/tasks/salesforce/UpdateDependencies.py", line 186, in _install_dependency
    headers=dependency.get("headers", {}),
  File "/Users/ant/.pyenv/versions/2.7.15/envs/cci/src/cumumulsci-0/cumulusci/utils.py", line 139, in download_extract_zip
    zip_file = zipfile.ZipFile(zip_content)
  File "/Users/ant/.pyenv/versions/2.7.15/lib/python2.7/zipfile.py", line 770, in __init__
    self._RealGetContents()
  File "/Users/ant/.pyenv/versions/2.7.15/lib/python2.7/zipfile.py", line 811, in _RealGetContents
    raise BadZipfile, "File is not a zip file"
zipfile.BadZipfile: File is not a zip file
```

**Comment by 1handclapping** [comment]
If I haven't got this completely messed up, here is [the changes](https://github.com/SFDO-Tooling/CumulusCI/commit/2e0be4e9ae7b6fdcab2e380995d5c766229d6d13) I made to get it working locally. It would seem that downloading the repo as a zip needs Basic Auth if it is private as the token auth in the header field was returning a 404.

Be more than enthusiastic to be guided towards a PR, be a first time for me! 😅

**Comment by cdcarter** [comment]
Hi @1handclapping! Thanks for your interest in CCI *and* in contributing back. This is a great find/edge case, we (SFDO) use private repositories with CCI, but none are dependencies at this time....

We actually fixed a bug similar to this on the web-based CI tool and switched to using a GitHub SDK instead of passing the auth headers ourselves. You can see the change here: https://github.com/SalesforceFoundation/MetaCI/commit/b29ba27b6f68a8e42928c1f9c5c0ef3453628736.

I think we need to do a similar thing here...instead of just calling download_extract_zip for everything, if its a named github dep at a version or ref, we should use the github api to get the zipball url. 

Let me know if that makes any sense, and we can also keep talking about the changes needed. Can you also email me (christian.carter@salesforce.com) so that we can get you set up with our contributor stuff?

**Comment by 1handclapping** [comment]
Hi @cdcarter! Ok great, I think I've got the idea. I'll start a new branch and give it a shot and report back shortly to firm up the approach. Awesome. 😄

**Comment by 1handclapping** [comment]
Where would the best place be to pick up the Github SDK instance to call the archive method? I'm currently passing it through as an item in the dictionaries in the repo_dependencies[] list from `process_github_dependency` but that feels a little out of place. However, to create a new instance at say `_install_dependency` in tasks.salesforce.UpdateDependencies you would have to duplicate the constructor call making it twice for each dep. In some ways, it would be nice to not have to re-authorize the SDK for each dependency at all?

**Comment by 1handclapping** [comment]
Thanks @davisagli, makes sense. My train of thought was in `process_github_dependency` where a repo instance is built to inspect contents:
```
# Initialize github3.py API against repo
        gh = self.get_github_api()
        repo_owner, repo_name = dependency["github"].split("/")[3:5]
        if repo_name.endswith(".git"):
            repo_name = repo_name[:-4]
        repo = gh.repository(repo_owner, repo_name)
```
Would the plan be to create a new repo instance in `_install_dependency` parsing out from the `zip_url` in the flattened deps list?. [At present I'm duplicating a lot of this](https://github.com/SFDO-Tooling/CumulusCI/compare/master...1handclapping:feature/support-private-repo) and passing the repo to `download_extract_github` utils method.  
  
There is an opportunity to pass the `repo` ref from `process_github_dependency` as an item in the dictionary but perhaps unnecessary?

As it stands I'm just testing for `https://github.com` in the zip_url at `_install_dependency` and thus any listed github dependency, either via github or zip_url in the yaml file will use the Github SDK. Is there a better way?

And another note on the github tag in the cumulusci.yml file. This is not passed on in the deps flattened list so currently can't get to it at `_install_dependency`. That led me to think, if I'm not mistaken, though whilst the tag is used to process the nested deps listed in the yaml file in `process_github_dependency`:
```
 # Determine the ref if specified
        kwargs = {}
        if "tag" in dependency:
            tag = dependency["tag"]
            kwargs["ref"] = tag
        else:
            tag = None

        # Get the cumulusci.yml file
        contents = repo.contents("cumulusci.yml", **kwargs)
        cumulusci_yml = hiyapyco.load(contents.decoded, loglevel="INFO")
```
the zip_url is built up using the `repo.default_branch` at 
```
zip_url = "{}/archive/{}.zip".format(repo.html_url, repo.default_branch)
```
and hence currently unpackaged source would always be at default branch regardless of if a specific tag was present in the yaml config? Not sure if that is expected behaviour.

TODO: 

- move the auth headers out in the `process_github_dependency` as they are no longer required.
- support github tag ref from cumulusci.yml

**Comment by davisagli** [comment]
@1handclapping oh...this is more complicated than I was remembering. I forgot that currently process_github_dependency returns dicts with a zip_url and then _install_dependency fetches that URL.

I like the plan of adding the repo to the dict returned from process_github_dependency, as well as the ref (which is often but not always a tag). And yeah, remove the zip_url and headers.

Then the logic in _install_dependency will be something like this:

```python
    package_zip = None
    if "zip_url" in dependency:
        self.logger.info(
            "Deploying unmanaged metadata from /{} of {}".format(
                dependency["subfolder"], dependency["zip_url"]
            )
        )
        package_zip = download_extract_zip(
            dependency["zip_url"],
            subfolder=dependency.get("subfolder"),
            headers=dependency.get("headers", {}),
        )
    elif 'repo' in dependency:
        # add a logger line here...
        package_zip = download_extract_github(
            dependency['repo'],
            dependency['ref'],
        )
    if package_zip:
        # ...do the rest of the processing...
```

You'll also want to modify BaseProjectConfig.pretty_dependencies to skip over the repo key instead of the headers key.

Sound reasonable?

**Comment by davisagli** [comment]
@1handclapping I skimmed through and it looks great so far. Once you're ready, open a pull request and we can do a more careful review, and also send you our contributor license agreement.

For unit testing, you can pass `download_extract_github` a mock object that pretends to be the github3 repository object:

```python
mock_repo = mock.Mock(default_branch='master')
zf = zipfile.ZipFile()
# ... put what you want in the mock zip file ...
mock_repo.archive.return_value = zf
result = download_extract_github(mock_repo, 'subfolder', 'ref')
# ... make assertions about the result ...
```

By the way, we are about to merge Python 3 support to master (keeping Python 2 support too for now). Don't worry about testing your changes in Python 3 unless you want to; I can make any necessary changes as we merge the pull request.

**Comment by 1handclapping** [comment]
@davisagli, thanks once again, and thats great news about py3. It turned out I had to use a `side_effect` on the archive mocked method as it expects the file like object to be passed in as an argument. Seems to be working now, sure we can discuss at review time.

I'm just finishing tidying up and running Black for the PR. I've also just noticed running black has added a bit of diff noise around adding dangling commas at the end of some lists?

Re the contrib guidelines, you need a new feature branch to accept the changes? I'll get this PR done tonight.


Thanks!

**Comment by davisagli** [comment]
I see what you mean about using `side_effect`. That looks okay.

I bet you are using a slightly different version of black than we have been running -- don't worry about it, I'll need to make a copy of your branch in order to run the tests with our CI system and I can run black on it then. Once we're doing development on Python 3 we can start including a version pin for black in requirements_dev.txt.

**Comment by 1handclapping** [comment]
ok sounds good. I'll wait to hear from you re the branch.

**Comment by davisagli** [comment]
It doesn't matter what the branch is called in your fork of the repository.

**Comment by 1handclapping** [comment]
ok right, sorry for the hand holding (fairly new for me on this front), I'll make the PR now. Sure there will be things to sort through.

**Comment by davisagli** [comment]
Released in 2.1.0
